staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go

```diff
@@ -490,3 +515,59 @@ func startDBSizeMonitorPerEndpoint(client *clientv3.Client, interval time.Durati
 		cancel()
 	}, nil
 }
+
+// checkVersion checks for deprecated versions of etcd that are affected by known issues.
+// It returns an error if the cluster is running a deprecated version, specifically versions below 3.4.25
+// or between 3.5.0 and 3.5.7. For more details on this issue, refer to:
+//   - https://github.com/etcd-io/etcd/issues/15220
+//   - https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/2340-Consistent-reads-from-cache/README.md#bug-in-etcd-progress-notification
+func checkVersion(c *clientv3.Client, dialTimeOut time.Duration) (err error) {
+	eps := c.Endpoints()
+	ctx := context.TODO()
+
+	if dialTimeOut > 0 {
+		var cancel context.CancelFunc
+		ctx, cancel = context.WithTimeout(ctx, dialTimeout)
+		defer cancel()
+	}
+	g, ctx := errgroup.WithContext(ctx)
+
+	for _, ep := range eps {
+		// if cluster is current, any endpoint gives a recent version
+		g.Go(checkVersionEndpoint(c, ctx, ep))
+	}
+	return g.Wait()
+}
+
+// isDeprecatedVersion checks if the given etcd version is affected by the known issue.
+// Versions below 3.4.25 or between 3.5.0 and 3.5.7 are considered deprecated.
+// Returns true if the version is deprecated, false otherwise.
+func isDeprecatedVersion(ver string) (bool, error) {
+	respVer, err := version.ParseSemantic(ver)
+	if err != nil {
+		return false, err
+	}
+	return respVer.LessThan(v3_4_25) || respVer.AtLeast(v3_5_0) && respVer.LessThan(v3_5_7), nil
+}
+
+// checkVersionEndpoint returns a closure that checks the version of the etcd cluster
```

Comment:
Don't need to return a function. Just have a function `func checkVersionEndpoint(c *clientv3.Client, ctx context.Context, endpoint string) error` and call it via closure:
```
	for i := range eps {
		ep := eps[i]
		g.Go(func()error {
			return checkVersionEndpoint(c, ctx, ep)
		})
	}
```

Reply:
That makes more sense, thanks for update